### ISSUE TITLE
Changing element order in materials changes simulation result

### ISSUE THREAD

**Issue by shimwell** [openmc-dev/openmc#1902]
Hi all,

I've been repeating the same simulation but with reordered elements in the material.xml files and noticed that the simulation results change.

I didn't purposefully reorder the elements in the material file, just the parameter runs tend to write out the materials in a different order from time to time.

Anyway with further investigation and production of a minimal example I am wondering if this is a bug in unified cross section creation or somewhere else.

Running a simple example with just two elements (lithium and lead) one way around and then the other order gives different tally results.

lead before lithium material
```xml
<?xml version='1.0' encoding='utf-8'?>
<materials>
  <material id="1" name="PbLi">
    <density units="atom/b-cm" value="0.032720171" />
    <nuclide ao="1.1788" name="Pb204" />
    <nuclide ao="20.2922" name="Pb206" />
    <nuclide ao="18.6082" name="Pb207" />
    <nuclide ao="44.1208" name="Pb208" />
    <nuclide ao="7.9" name="Li6" />
    <nuclide ao="7.9" name="Li7" />
  </material>
</materials>
```

lithium before lead material
```xml
<materials>
  <material id="1" name="PbLi">
    <density units="atom/b-cm" value="0.032720171" />
    <nuclide ao="7.9" name="Li6" />
    <nuclide ao="7.9" name="Li7" />
    <nuclide ao="1.1788" name="Pb204" />
    <nuclide ao="20.2922" name="Pb206" />
    <nuclide ao="18.6082" name="Pb207" />
    <nuclide ao="44.1208" name="Pb208" />
  </material>
</materials>
```

Runs with lead appear in the materials.xml file first
1.684769278870781
1.6847692788707778
1.6847692788707815
1.6847692788707818
1.6847692788707809
1.6847692788707795
1.6847692788707818
1.6847692788707798
1.6847692788707829
1.6847692788707815

Runs with lithium appearing in the materials.xml file first
1.6851566043500135
1.6851566043500101
1.6851566043500101
1.685156604350011
1.6851566043500106
1.685156604350012
1.6851566043500108
1.6851566043500106
1.6851566043500095
1.6851566043500157

While the repeats of the same materials.xml, geometry.xml files are fairly constant. There appears to be a step change when the order of materials is switched in the material.xml
1.685... and 1.684

Perhaps this is related to the unordered materials discussed here #1813 

I attach the openmc simulation files for reproduction and am using version 0.13.0
Git SHA1: f0c1ce8bb138767baaaad8ec19c21a6b835972be on Ubuntu 18.04

[lithium_before_lead.zip](https://github.com/openmc-dev/openmc/files/7433353/lithium_before_lead.zip)

[lead_before_lithium.zip](https://github.com/openmc-dev/openmc/files/7433354/lead_before_lithium.zip)

**Comment by shimwell** [comment]
yep that sounds right to me and naturally the difference can be removed whem particles are increased.

However I'm still going to have to either sort or randomize the material order in the materials.xml file.
Otherwise parameter studies will not give a smooth surface. Which makes optimization harder than it needs to be.

Would people be open to a PR that sorts the material.xml file by ZAID number?

**Comment by gridley** [comment]
> However I'm still going to have to either sort or fully randomize the material order in the materials.xml file. Otherwise parameter studies will not give a smooth surface. Which makes optimization harder than it needs to be.

I am 99% sure you will not get a smooth surface regardless, unless you are running a problem with no scattering or something like that. Once you slightly perturb a problem, particles are most likely going to take entirely different tracks, not simple perturbations from the original ones.

> Would people be open to a PR that sorts the material.xml file by ZAID number?

My personal opinion is that we should sort the nuclide list in C++ after reading. The output of the program should be independent of a physically (but not computationally) identical input. Then nobody would have to worry about calling some python API method to sort the nuclides ever again.

On the other hand, sorting them in C++ would likely require re-golding our regression tests, unless by some strange coincidence the nuclides in the tests are sorted already.

**Comment by paulromano** [comment]
Agree with @gridley's comment on optimization -- I don't see how you would get a smooth surface. Even if you run the same problem twice on the same machine, if it is done in parallel it's entirely possible you'll end up with two different but statically equivalent answers.

Personally, I am not in favor of doing any sorting by default unless there is a good reason for it. The only "good" reason I can think of is to improve the efficiency of sampling a nuclide for a collision (which would have to be based on some combination of the atom density of that nuclide along with its cross section), but even for that I'd need some clear evidence that it's worth the extra complexity. In the absence of sorting by default, it should be sufficient to provide a method / means by which a user can sort nuclides in materials if they so choose (`Materials.sort_nuclides()` or something like that).

**Comment by gridley** [comment]
@paulromano there is a benefit to sorting nuclides for GPU operations. Imagine you have two  fuel mixtures, one with nuclides listed in reverse order. Now consider you're in the fuel XS lookup kernel, and all even-numbered particles are in fuel mix 1 and the others in fuel mix 2. You are likely to use whatever's in the cache better if there is a chance the whole thread block is looking at the same nuclide. Moreover, with sorted nuclides, you could sync threads after each nuclide. I have heard through the grapevine that a certain GPU MC code that's about 10 times faster than GPU SHIFT uses this technique...
